text/0000-editions.md

```diff
@@ -0,0 +1,86 @@
+- Start Date: 2018-09-04
+- RFC PR:
+- Ember Issue:
+
+# Editions
+
+## Summary
+
+Introduce a new concept: editions. Every few years, we will declare a new edition of Ember that bundles up accumulated incremental improvements into a cohesive package.
+
+Editions give us an opportunity to bring our documentation and marketing up-to-date to reflect the improvements we’ve made since the previous edition. They are also the time to synchronize Ember and the wider ecosystem, including updating default settings, blueprints, addons, and more.
+
+Editions also help ensure that Ember users are getting a cohesive, polished experience. The right time to declare a new edition is when:
+
+
+- A significant, coherent set of new features and APIs have all landed in the stable channel.
+- Error messages and the developer ergonomics of those new features have been fully polished.
+- Tooling (the Ember Inspector, blueprints, codemods, etc.) has been updated to work with these new features.
+- API documentation, guides, tutorials, and example code has been updated to incorporate the new features.
+
+## Motivation
+
+Ember releases a new, stable version every six weeks. For existing users, this drumbeat of incremental improvement is easier to keep up with than splashy, big-bang releases.
+
+However, this steady stream of small improvements comes with some downsides.
+
+- For Ember users, it’s hard to keep track of evolving best practices, and understand how new features are supposed to work together.
+- For non-Ember users, it’s not obvious what has changed and whether Ember is worth another look.
+- For Ember contributors, it’s not clear when to update documentation and marketing, because there’s always more features in the pipeline.
+
+This RFC proposes a new release concept called “editions”, which provides a lens through which to view these accumulated changes. It also marks an opportunity to synchronize the wider ecosystem with the latest changes to Ember, including updates to defaults, documentation, addons, and more.
+
+## Detailed design
+
+Editions touch several parts of the Ember ecosystem, and generally tie together a set of incremental changes with a unifying theme. Most of these changes should come from existing community processes (i.e. RFCs), with editions serving as a higher level conceptual grouping.
+
+### Editions vs. Releases
+Editions are not separate releases, but rather a specific range of releases belong to an edition. The current edition will be open-ended - it’s terminating release is determined by whenever the next edition is declared.
+
+### Proposing Editions
+Each edition is proposed via a RFC, which should describe in detail the change in programming model and how to communicate that with users.
```

Comment:
> programming model

Would you mind unpacking this term a bit?

I'm used to thinking of "programming model" as a low-level concept tied to whatever the program is running on (i.e., all languages that run on the JVM or BEAM share a programming model, despite having very different syntaxes and APIs).

Reply:
I think @tomdale likely has the most articulate version of this in his head, but I'll take a stab.

Ember's features tend to land incrementally and piecemeal. This makes for an easier upgrade path, and ensures that new functionality is backwards compatible. 

But sometimes, a set of separate, individual features can work together to produce a developer experience that is greater than the sum of the parts. For example, take angle bracket components, Glimmer, module unification, and some of the other features currently in the pipeline. Taken together, they create a vastly different experience for developers building with Ember.

But because of the incremental progress made to land these features, they might land in different versions, and will almost certainly not land in a major version (i.e. 4.0). So when a minor version ships with just one of these features - do we update the docs? Do we start pushing it as best practice?

If we do, the docs & overall developer experience might feel incongruent (because the one feature was designed to make the most sense with the other features in mind, even though it's backwards compatible). If we don't, when does experience shift? If the answer is "after all the related features ship", then that's basically what an edition is designed to accomplish.

Editions gives us a chance to introduce a higher level abstraction that operates similar to the way most other JS projects use major versions: they mark paradigm shifts in how you use the framework. But because Ember reserves major versions for the strictly technical purpose of indicating the removal of deprecated APIs, we lack such a tool. Editions aim to fill that gap.

Hopefully that clarifies a bit. If not, I'll defer to @tomdale's more articulate writing skills.

Reply:
👍 to what @davewasmer said. When I use "programming model," I'm talking about the mental model developers should have of the framework and how different abstractions can be combined can be deployed to solve a particular problem.

I think @davewasmer's description of a set of features producing "greater than the sum of the parts" is spot on. For example, the Ember 1.x programming model was driven by two-way bindings. The Ember 2.x programming model is driven more by DDAU. But I think we've done a poor job of communicating that that shift has happened, so it's not always clear to developers when they should use two-way bindings vs. `mut` helpers vs. passing functions, etc. Bubbling vs. closure actions is another example of where we've allowed the programming model to be muddied.

A big part of the goal of editions is to serve as a discrete moment where we can have a discussion as a community about what our shared best practices are.